allennlp/data/iterators/bucket_iterator.py

```diff
@@ -164,3 +127,51 @@ def _create_batches(self, instances: Iterable[Instance], shuffle: bool) -> Itera
                 batches.insert(0, last_batch)
 
             yield from batches
+
+    def _sort_by_padding(self, instances: List[Instance]) -> List[Instance]:
+        """
+        Sorts the instances by their padding lengths, using the keys in
+        ``sorting_keys`` (in the order in which they are provided).  ``sorting_keys`` is a list of
+        ``(field_name, padding_key)`` tuples.
+        """
+        if not self._sorting_keys:
+            self._guess_sorting_keys(instances)
+        instances_with_lengths = []
+        for instance in instances:
+            # Make sure instance is indexed before calling .get_padding
+            instance.index_fields(self.vocab)
+            padding_lengths = cast(Dict[str, Dict[str, float]], instance.get_padding_lengths())
+            if self._padding_noise > 0.0:
+                noisy_lengths = {}
+                for field_name, field_lengths in padding_lengths.items():
+                    noisy_lengths[field_name] = add_noise_to_dict_values(
+                        field_lengths, self._padding_noise
+                    )
+                padding_lengths = noisy_lengths
+            instance_with_lengths = (
+                [
+                    padding_lengths[field_name][padding_key]
+                    for (field_name, padding_key) in self._sorting_keys
+                ],
+                instance,
+            )
+            instances_with_lengths.append(instance_with_lengths)
+        instances_with_lengths.sort(key=lambda x: x[0])
+        return [instance_with_lengths[-1] for instance_with_lengths in instances_with_lengths]
+
+    def _guess_sorting_key(self, instances: List[Instance]) -> None:
+        max_length = 0
+        longest_padding_key: Tuple[str, str] = None
+        for instance in instances:
+            instance.index_fields(self.vocab)
+            padding_lengths = cast(Dict[str, Dict[str, float]], instance.get_padding_lengths())
+            for field_name, field_padding in padding_lengths.items():
+                for padding_key, length in field_padding.items():
+                    if length > max_length:
+                        max_length = length
+                        longest_padding_key = (field_name, padding_key)
+        if not longest_padding_key:
+            raise ValueError(
+                "Found no field that needed padding; why are you using a bucket iterator?"
```

Comment:
This should probably be a `ConfgigurationError`, as it explicitly relates to a poorly configured config. Also, make the error message better. This one assumes knowledge - instead, advise what someone should use, rather than just telling them they're doing something wrong.

Reply:
I made it an assertion error instead, with a direction to open an issue on github; it's basically impossible to get there, and it means I made a wrong assumption somewhere in the logic.